pkg/ext/stores/useractivity/store.go

```diff
@@ -0,0 +1,307 @@
+func (s *Store) GroupVersionKind(_ schema.GroupVersion) schema.GroupVersionKind {
+	return GVK
+}
+
+// NamespaceScoped implements [rest.Scoper]
+func (s *Store) NamespaceScoped() bool {
+	return false
+}
+
+// GetSingularName implements [rest.SingularNameProvider]
+func (s *Store) GetSingularName() string {
+	return SingularName
+}
+
+// New implements [rest.Storage]
+func (s *Store) New() runtime.Object {
+	obj := &ext.UserActivity{}
+	obj.GetObjectKind().SetGroupVersionKind(GVK)
+	return obj
+}
+
+// Destroy implements [rest.Storage]
+func (s *Store) Destroy() {
+}
+
+// Create implements [rest.Creator]
+// Create sets the Status fields on the UserActivity object
+// provided by the user within the request.
+func (s *Store) Create(ctx context.Context,
+	obj runtime.Object,
+	createValidation rest.ValidateObjectFunc,
+	options *metav1.CreateOptions) (runtime.Object, error) {
+	userInfo, err := s.userFrom(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	extras := userInfo.GetExtra()
+
+	authTokenID := first(extras[common.ExtraRequestTokenID])
+	if authTokenID == "" {
+		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("missing request token ID"))
+	}
+
+	if createValidation != nil {
+		err := createValidation(ctx, obj)
+		if err != nil {
+			return obj, err
+		}
+	}
+
+	// retrieving useractivity object from raw data
+	objUserActivity, ok := obj.(*ext.UserActivity)
+	if !ok {
+		var zeroUA *ext.UserActivity
+		return nil, apierrors.NewInternalError(fmt.Errorf("expected %T but got %T", zeroUA, objUserActivity))
+	}
+	// retrieve token information
+	if objUserActivity.Name == "" {
+		return nil, apierrors.NewBadRequest("can't retrieve token with empty string")
+	}
+
+	// retrieve auth token
+	authToken, err := s.tokenCache.Get(authTokenID)
+	if err != nil {
+		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("error getting request token %s: %w", authTokenID, err))
+	}
+
+	// retrieve activity token
+	activityToken, err := s.tokenCache.Get(objUserActivity.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, apierrors.NewBadRequest(fmt.Sprintf("token not found %s: %v", objUserActivity.Name, err))
+		} else {
+			return nil, apierrors.NewInternalError(fmt.Errorf("failed to get token %s: %w", objUserActivity.Name, err))
+		}
+	}
+
+	// verify auth and activity token has the same auth provider
+	if authToken.AuthProvider != activityToken.AuthProvider {
+		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("request token %s and activity token %s have different auth providers", authTokenID, objUserActivity.Name))
+	}
+
+	// verify auth and activity token has the same auth user principal
+	if authToken.UserPrincipal.Name != activityToken.UserPrincipal.Name {
+		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("request token %s and activity token %s have different user principals", authTokenID, objUserActivity.Name))
+	}
+
```

Comment:
If I understand correctly user activity is only applicable to session tokens. Therefore we have to make sure that token we're about to update is in fact is a session token (the label `authn.management.cattle.io/kind` is set to `session`).

Reply:
that's correct, thanks for the advice